### DISCUSSION TITLE
Get better clone performance by replacing `git clone` with `scalar clone`

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#18329] [Suggest an Idea]
Git `2.38` comes with _Scalar_ [^gh-blog], a new built-in repository management tool. It sounds like Scalar is designed to be fast when cloning large repositories.

Thoughts:

- Replacing `git clone` with `scalar clone` may speed up repository clones for Renovate bot.
- `scalar clone` may be easier to use than `git clone`, as Scalar comes with default settings.

I'll let you guys look into this to see if Scalar is a good fit for Renovate. 😉 

[^gh-blog]: https://github.blog/2022-10-03-highlights-from-git-2-38/#a-repository-management-tool-for-large-repositories

**Comment by rarkins** [comment]
"comes with default settings" is not a reason to convince me. We depend so heavily on git, we should know what settings we're using. I think it's targeted at end users who don't want to understand all the ways to configure and just want good defaults.

So the question is: does it do things which `git clone` cannot, which would speed things up for us?

**Reply by HonkingGoose** [reply]
## Summary

New features in `scalar`:

1. automatically configuring built-in Git features
1. managing repository enlistments

You don't care about "comes with default settings", so the only thing left is the _enlistment_.

I think enlistments is designed to nudge people to put their build-artifacts _outside_ the `src` directory. This way they avoid the performance hit of git-ignoring their build-artifacts.

Because the default-behavior is baked into Scalar, we may run into problems when running an old version of `scalar clone` against an user's repository. Because Scalar is built-in to Git now, we should explicitly list the Git version support in our program, if we don't already.

## Git-scm on Scalar features [^git-scm-features]

> **Features**
> 
> Scalar is comprised of two major pieces of functionality: automatically configuring built-in Git performance features and managing repository enlistments.
> 
> The Git performance features configured by Scalar (see "Background" for examples) confer substantial performance benefits to large repositories, but are either too experimental to enable for all of Git yet, or only benefit large repositories. As new features are introduced, Scalar should be updated accordingly to incorporate them. This will prevent the tool from becoming stale while also providing a path for more easily bringing features to the appropriate users.
> 
> Enlistments are how Scalar knows which repositories on a user’s system should utilize Scalar-configured features. This allows it to update performance settings when new ones are added to the tool, as well as centrally manage repository maintenance. The enlistment structure - a root directory with a `src/` subdirectory containing the cloned repository itself - is designed to encourage users to route build outputs outside of the repository to avoid the performance-limiting overhead of ignoring those files in Git.

## Quotes from Scalar manual

I picked some relevant quotes from Git's Scalar manual. [^manual]

### Official goal of Scalar

> Scalar is a repository management tool that optimizes Git for use in large
> repositories. Scalar improves performance by configuring advanced Git settings,
> maintaining repositories in the background, and helping to reduce data sent
> across the network.

### Enlistment

> An important Scalar concept is the enlistment: this is the top-level directory of the project.
> It usually contains the subdirectory `src/` which is a Git worktree.
> This encourages the separation between tracked files (inside `src/`) and untracked files, such as build artifacts (outside `src/`).
> When registering an existing Git worktree with Scalar whose name is not `src`, the enlistment will be identical to the worktree.

### Run `scalar reconfigure` after each update!

Big gotcha here, you need to run `scalar reconfigure` after updating the Scalar program:

> Reconfigure
> ---
> 
> After a Scalar upgrade, or when the configuration of a Scalar enlistment
> was somehow corrupted or changed by mistake, this subcommand allows to
> reconfigure the enlistment.
> 
> With the `--all` option, all enlistments currently registered with Scalar
> will be reconfigured. Use this option after each Scalar upgrade.

### Diagnosing problems with Scalar

> Diagnose
> ---
> 
> `diagnose [<enlistment>]`:
>     When reporting issues with Scalar, it is often helpful to provide the information gathered by this command, including logs and certain statistics describing the data shape of the current enlistment.
> 
> The output of this command is a `.zip` file that is written into a directory adjacent to the worktree in the `src` directory.

[^git-scm-features]: [Scalar docs on git-scm](https://git-scm.com/docs/scalar)
[^manual]: [Git manpage for `scalar` for version `v2.38.0`](https://github.com/git/git/blob/v2.38.0/Documentation/scalar.txt)
